assets/src/edit-story/components/checklist/checks/videoOptimization.js

```diff
@@ -0,0 +1,199 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import { _n, __ } from '@web-stories-wp/i18n';
+import styled from 'styled-components';
+
+/**
+ * Internal dependencies
+ */
+import { useCallback, useMemo, useRef } from 'react';
+import { useLocalMedia, useStory } from '../../../app';
+import { VIDEO_SIZE_THRESHOLD } from '../../../app/media/utils/useFFmpeg';
+import { PRIORITY_COPY } from '../constants';
+import { LayerThumbnail, Thumbnail, THUMBNAIL_TYPES } from '../../thumbnail';
+import {
+  CARD_TYPE,
+  ChecklistCard,
+  DefaultFooterText,
+} from '../../checklistCard';
+import { filterStoryElements } from '../utils/filterStoryElements';
+import {
+  Button,
+  BUTTON_SIZES,
+  BUTTON_TYPES,
+  Tooltip,
+} from '../../../../design-system';
+import { useHighlights } from '../../../app/highlights';
+import { useRegisterCheck } from '../checkCountContext';
+import { StyledVideoOptimizationIcon } from '../../checklistCard/styles';
+
+const OptimizeButton = styled(Button)`
+  margin-top: 4px;
+`;
+
+export function isVideoElementOptimized(element = {}) {
+  if (element.resource?.isTranscoding) {
+    return true;
+  }
+
+  if (
+    element.type !== 'video' ||
+    element.resource?.local ||
+    element.resource?.isOptimized
+  ) {
+    return false;
+  }
+
+  const videoArea =
+    (element.resource?.height ?? 0) * (element.resource?.width ?? 0);
+  const isLargeVideo =
+    videoArea >= VIDEO_SIZE_THRESHOLD.WIDTH * VIDEO_SIZE_THRESHOLD.HEIGHT;
+  return isLargeVideo;
+}
+
+export const BulkVideoOptimization = () => {
+  const story = useStory(({ state }) => state);
+  const optimizingResourcesRef = useRef({});
+  const unoptimizedVideos = filterStoryElements(story, isVideoElementOptimized);
+
+  const setHighlights = useHighlights(({ setHighlights }) => setHighlights);
+
+  const { optimizeVideo } = useLocalMedia((state) => ({
+    optimizeVideo: state.actions.optimizeVideo,
+  }));
+
+  const sequencedVideoOptimization = useCallback(() => {
+    // only attempt to optimize videos that are not currently being transcoded
+    const optimizingResources = unoptimizedVideos.filter(
+      ({ resource }) => !resource?.isTranscoding
+    );
+    return new Promise((res) => {
+      optimizingResources
+        .map(({ resource }) => async () => {
+          await optimizeVideo({ resource }).finally(() => {
+            optimizingResourcesRef.current[resource.id] = 'UPLOADED';
```

Comment:
nit: add `UPLOADED` and `UPLOADING` as constants since you are using those a few places here.